mirascope/mcp/client.py

```diff
@@ -0,0 +1,280 @@
+        return build_object_type(properties, required, name)
+    else:
+        # Default fallback
+        return Any
+
+
+def create_model_from_tool(tool: Tool) -> type[BaseModel]:
+    """
+    Create a dynamic pydantic model from the given Tool instance based on its inputSchema.
+    """
+    schema = tool.inputSchema
+    properties = schema.get("properties", {})
+    required_fields = schema.get("required", [])
+
+    fields = {}
+    for field_name, field_schema in properties.items():
+        field_type = json_schema_to_python_type(field_schema)
+
+        if field_name in required_fields:
+            default = Field(..., description=field_schema.get("description"))
+            annotation = field_type
+        else:
+            default = Field(None, description=field_schema.get("description"))
+            annotation = field_type | None
+
+        fields[field_name] = (annotation, default)
+
+    return create_model(
+        snake_to_pascal(tool.name), __config__=ConfigDict(extra="allow"), **fields
+    )
+
+
+def _create_model_from_prompt(prompt: Prompt) -> type[BaseModel]:
+    if not prompt.arguments:
+        return create_model(prompt.name, __base__=MCPClientTool)
+
+    fields = {}
+    for arg in prompt.arguments:
+        if arg.required:
+            default = Field(..., description=arg.description)
+            annotation = str
+        else:
+            default = Field(None, description=arg.description)
+            annotation = str | None
+
+        fields[arg.name] = annotation, default
+
+    return create_model(prompt.name, __base__=MCPClientTool, **fields)
+
+
+class MCPClient:
+    def __init__(self, session: ClientSession) -> None:
+        self.session: ClientSession = session
+
+    async def list_resources(self) -> list[Resource]:
+        result = await self.session.list_resources()
+        return result.resources
+
+    async def read_resource(
+        self, uri: str | AnyUrl
+    ) -> list[str | BlobResourceContents]:
+        result = await self.session.read_resource(
+            uri if isinstance(uri, AnyUrl) else uri
+        )
+        parsed_results = []
+        for content in result.contents:
+            if isinstance(result, BlobResourceContents):
+                parsed_results.append(result)
+            else:
+                parsed_results.append(content.text)
+        return parsed_results
+
+    async def list_prompts(self) -> list[Prompt]:
+        result = await self.session.list_prompts()
+        return result.prompts
+
+    async def get_prompt_template(
+        self, name: str
+    ) -> Callable[..., Awaitable[list[BaseMessageParam]]]:
+        async def async_prompt(**kwargs: str) -> list[BaseMessageParam]:
+            result = await self.session.get_prompt(name, kwargs)
+
+            return [
+                _convert_prompt_message_to_base_message_params(prompt_message)
+                for prompt_message in result.messages
+            ]
+
+        return async_prompt
```

Comment:
In this method, create a async function `Callable[..., Awaitable[list[BaseMessageParam]]]`